packages/agent/test/agent_spec.ts

```diff
@@ -0,0 +1,181 @@
+      proxyUrl: `http://foo:bar@localhost:${PROXY_PORT}`,
+      proxyAuth: true,
+    },
+    {
+      name: 'With an HTTPS upstream requiring auth',
+      proxyUrl: `https://foo:bar@localhost:${PROXY_PORT}`,
+      httpsProxy: true,
+      proxyAuth: true
+    }
+  ].slice().map((testCase) => {
+    context(testCase.name, function() {
+      beforeEach(function() {
+        this.oldEnv = Object.assign({}, process.env)
+        process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'
+
+        if (testCase.proxyUrl) {
+          process.env.HTTP_PROXY = process.env.HTTPS_PROXY = testCase.proxyUrl
+          process.env.NO_PROXY = ''
+        }
+
+        this.agent = new CombinedAgent()
+
+        this.request = request.defaults({
+          proxy: null,
+          strictSSL: false,
+          agent: this.agent
+        })
+
+        if (testCase.proxyUrl) {
+          let options: any = {
+            keepRequests: true,
+            https: false,
+            auth: false
+          }
+
+          if (testCase.httpsProxy) {
+            options.https = this.https
+          }
+
+          if (testCase.proxyAuth) {
+            options.auth = {
+              username: 'foo',
+              password: 'bar'
+            }
+          }
+
+          this.debugProxy = new DebuggingProxy(options)
+          return this.debugProxy.start(PROXY_PORT)
+        }
+      })
+
+      afterEach(function() {
+        process.env = this.oldEnv
+
+        if (testCase.proxyUrl) {
+          this.debugProxy.stop()
+        }
+      })
+
+      it('HTTP pages can be loaded', function() {
+        return this.request({
+          url: `http://localhost:${HTTP_PORT}/get`,
+        }).then(body => {
+          expect(body).to.eq('It worked!')
+          if (this.debugProxy) {
+            expect(this.debugProxy.requests[0]).to.include({
+              url: `http://localhost:${HTTP_PORT}/get`
+            })
+          }
+        })
+      })
+
+      it('HTTPS pages can be loaded', function() {
+        return this.request({
+          url: `https://localhost:${HTTPS_PORT}/get`
+        }).then(body => {
+          expect(body).to.eq('It worked!')
+          if (this.debugProxy) {
+            expect(this.debugProxy.requests[0]).to.include({
+              https: true,
+              url: `localhost:${HTTPS_PORT}`
+            })
+          }
+        })
+      })
+
+      it('HTTP errors are catchable', function() {
+        return this.request({
+          url: `http://localhost:${HTTP_PORT}/empty-response`,
+        }).catch(err => {
+          if (this.debugProxy) {
+            expect(this.debugProxy.requests[0]).to.include({
+              url: `http://localhost:${HTTP_PORT}/empty-response`
+            })
+            expect(err.statusCode).to.eq(502)
+          } else {
+            expect(err.message).to.eq('Error: socket hang up')
+          }
+        })
+      })
+
+      it('HTTPS errors are catchable', function() {
```

Comment:
Where are the tests that prove we are closing down the sockets instead of instantiating them and having them grow over time?

I believe we need tests / examples of the client, the origin server, or the upstream proxy all destroying the sockets or disconnecting naturally - which enables us to observe and test the underlying agent / socket pooling behavior.

Would love to have tests that reproduce all the errors that fire in a way that's real and not contrived.

Reply:
I wasn't sure how to test this, since the behavior of having the sockets stay alive for 2 minutes before closing them is desirable, but we shouldn't have a test that just waits 2 minutes after making requests and then checks if sockets are open. I'm open to ideas.